### PR TITLE
[LLD][COFF] Add support for ARM64EC entry thunks.

### PR DESCRIPTION
**Pull request by cjacek**
For x86_64 callable functions, ARM64EC requires an entry thunk generated by the compiler. The linker interprets .hybmp sections to associate function chunks with their entry points and writes an offset to thunks preceding function section contents.

Additionally, ICF needs to be aware of entry thunks to not consider chunks to be equal when they have different entry thunks, and GC needs to mark entry thunks together with function chunks.

I used a new SectionChunkEC class instead of storing entry thunks in SectionChunk, following the guideline to keep SectionChunk as compact as possible. This way, there is no memory usage increase on non-EC targets.

### CODE REVIEW

**Comment by cjacek** [comment]
CC @bylaws

File: lld/COFF/InputFiles.cpp [hunk 1]
```diff
@@ -260,6 +297,8 @@ SectionChunk *ObjFile::readSection(uint32_t sectionNumber,
     guardEHContChunks.push_back(c);
   else if (name == ".sxdata")
     sxDataChunks.push_back(c);
+  else if (isArm64EC(getMachineType()) && name == ".hybmp$x")
```
Comment (mstorsjo) [inline lld/COFF/InputFiles.cpp:300]: Will the section name always have the `$x` suffix, or should we have something like `name.substr(0, name.find('$')) == ".hybmp"`?

File: llvm/include/llvm/BinaryFormat/COFF.h [hunk 2]
```diff
@@ -806,7 +806,7 @@ enum Feat00Flags : uint32_t {
   Kernel = 0x40000000,
 };
 
-enum class Arm64ECThunkType : uint8_t {
+enum Arm64ECThunkType : uint8_t {
```
Comment (mstorsjo) [inline llvm/include/llvm/BinaryFormat/COFF.h:809]: Why is this change needed here?

File: lld/test/COFF/arm64ec-entry-thunk.s [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+    .globl thunk
+    .p2align 2
+thunk:
+    mov w0, #10
+    ret
+
+    .section .hybmp$x, "yi"
+    .symidx func
+    .symidx thunk
+    .word 1
```
Comment (mstorsjo) [inline lld/test/COFF/arm64ec-entry-thunk.s:24]: A comment with what the `.word 1` means (the thunk type) would be good, maybe just `# entry thunk` on the same line.

File: lld/COFF/Chunks.cpp [hunk 4]
```diff
@@ -410,6 +410,10 @@ void SectionChunk::writeTo(uint8_t *buf) const {
 
     applyRelocation(buf + rel.VirtualAddress, rel);
   }
+
+  // Write the offset to EC entry thunk preceding section contents.
+  if (Defined *entryThunk = getEntryThunk())
+    write32le(buf - sizeof(uint32_t), entryThunk->getRVA() - rva + 1);
```
Comment (mstorsjo) [inline lld/COFF/Chunks.cpp:416]: The `+1` feels quite odd here. I don't doubt that it's right, but it feels quite surprising to see an odd number when it comes to arm64 function offsets (which all should be multiples of 4). Is there something more you can write here to explain it?

File: lld/COFF/InputFiles.cpp [hunk 5]
```diff
@@ -260,6 +297,8 @@ SectionChunk *ObjFile::readSection(uint32_t sectionNumber,
...
+    hybmpChunks.push_back(c);
```
Comment (mstorsjo) [inline lld/COFF/InputFiles.cpp:301]: So, this change both has the effect that the `.hybmp` section chunks are getting parsed, but also means that they are skipped and not included in the output file?

File: lld/test/COFF/arm64ec-entry-thunk.s [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+// RUN: llvm-mc -filetype=obj -triple=arm64ec-windows test-split-thunk.s -o test-split-thunk.obj
+// RUN: llvm-mc -filetype=obj -triple=arm64ec-windows test-split-hybmp.s -o test-split-hybmp.obj
+// RUN: llvm-mc -filetype=obj -triple=arm64ec-windows test-split-data.s -o test-split-data.obj
+// RUN: lld-link -machine:arm64ec -dll -noentry -out:out-split.dll loadcfg.obj \
+// RUN:          test-split-func.obj test-split-thunk.obj test-split-data.obj test-split-hybmp.obj
+// RUN: llvm-objdump -d out-split.dll | FileCheck --check-prefix=DISASM %s
+
+#--- test-align.s
+// Build multiple functions with thunks and various alignments and check that entry thunk offsets
+// are correctly placed.
```
Comment (mstorsjo) [inline lld/test/COFF/arm64ec-entry-thunk.s:87]: I don't see where the extra alignment actually is visible in the output - it seems like those cases end up perfectly aligned out of the box, so there's no padding visible anywhere in the disassembled text section?

File: lld/test/COFF/arm64ec-hybmp.s [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    ret
+
+    .section .hybmp$x,"yi"
+    .symidx func
+    .symidx thunk
+    .word 1
+
+// RUN: llvm-mc -filetype=obj -triple=arm64ec-windows text-func.s -o text-func.obj
+// RUN: not lld-link -machine:arm64ec -dll -noentry -out:test.dll text-func.obj 2>&1 | FileCheck -check-prefix=FUNC-NON-COMDAT %s
+// FUNC-NON-COMDAT: error: non COMDAT symbol 'func' in hybrid map
```
Comment (mstorsjo) [inline lld/test/COFF/arm64ec-hybmp.s:25]: I'm curious - why is it relevant for the entry thunk that the function is a COMDAT symbol? (Edit: After looking at more of the code, I realize that this is because we need space to add the offset value before - for non-COMDAT, the section chunk contains multiple functions, and we can't inject the offset right before the function.)

File: lld/test/COFF/arm64ec-hybmp.s [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+#--- offset-func.s
+    .section .text,"xr",discard,func
+    mov w0, #2
+    .globl func
+    .p2align 2, 0x0
+func:
+    mov w0, #1
```
Comment (mstorsjo) [inline lld/test/COFF/arm64ec-hybmp.s:33]: It took me a while to spot what was wrong here - that the function label isn't at the start of the section. It could be good to have a comment above this test saying what it tests (even if "offset" kinda gives it away, but it's not very obvious).

It could also be good to spell out a comment in the source, that we also check that the symbol points directly at the start of the section chunk.

**Review by mstorsjo — COMMENTED** [review]
Looks quite good to me overall. A handful of minor requests for more comments or general discussion.

File: lld/COFF/InputFiles.cpp (hunk 1 above)
Comment (cjacek) [reply lld/COFF/InputFiles.cpp:300]: I checked with MSVC, it seems to apply the special treatment only for that exact name. Things like ".hybmp", ".hybmp$xy" or ".hybmp$y" don't get that treatment.

File: lld/COFF/InputFiles.cpp [hunk 9]
```diff
@@ -152,6 +152,38 @@ void ObjFile::parseLazy() {
...
+  ulittle32_t src;
+  ulittle32_t dst;
+  ulittle32_t type;
+};
+
+void ObjFile::initializeECThunks() {
+  for (SectionChunk *chunk : hybmpChunks) {
+    if (chunk->getContents().size() % sizeof(ECMapEntry)) {
+      error("Invalid .hybmp chunk size " + Twine(chunk->getContents().size()));
+      return;
```
[the remark this answers is not in the document]
Comment (cjacek) [reply lld/COFF/InputFiles.cpp]: Good point, I changed it.

File: llvm/include/llvm/BinaryFormat/COFF.h (hunk 2 above)
Comment (cjacek) [reply llvm/include/llvm/BinaryFormat/COFF.h:809]: It's to make switch statement in `initializeECThunks` possible. Otherwise, there is no implicit conversion from `ulittle32_t` and the switch statement would need an explicit cast. And with that cast, we wouldn't be able to use `default` far the warning. I could probably make it work, but the main motivation for having it in a public header was LLD (in #85936), changing it here felt right.

File: lld/COFF/Chunks.cpp (hunk 4 above)
Comment (cjacek) [reply lld/COFF/Chunks.cpp:418]: It's what I observed with MSVC and what Windows expects, I can only speculate about the reason. It could be just "have the low bit set", or it could be interpreted as the offset being from the last byte of the padding, not the first byte of the function. I added a comment about it.

File: lld/COFF/InputFiles.cpp (hunk 5 above)
Comment (cjacek) [reply lld/COFF/InputFiles.cpp:301]: Yes, there is no reason for them to be in the output file. Entry thunks are looked up from the offset in padding in runtime. Later, the linker will need exit thunk association for generating import thunks, but once the image is linked, there is no need for such map.

File: lld/test/COFF/arm64ec-entry-thunk.s (hunk 6 above)
Comment (cjacek) [reply lld/test/COFF/arm64ec-entry-thunk.s:87]: Oh, right, it's missing cases where additional padding is required, I added more tests. (It's still good to make sure that we correctly handle padding that happens to match alignment).

File: lld/test/COFF/arm64ec-hybmp.s (hunk 8 above)
Comment (cjacek) [reply lld/test/COFF/arm64ec-hybmp.s:34]: I added a comment both here and in the code. It also clarifies why COMDATs are needed and adds more comments to tests.

Thanks for the review!

File: lld/COFF/Chunks.cpp [hunk 10]
```diff
@@ -411,7 +411,9 @@ void SectionChunk::writeTo(uint8_t *buf) const {
     applyRelocation(buf + rel.VirtualAddress, rel);
   }
 
-  // Write the offset to EC entry thunk preceding section contents.
+  // Write the offset to EC entry thunk preceding section contents. The low bit
+  // is always set, so it's effectively an offset from the last byte of the
+  // padding.
```
Comment (mstorsjo) [inline lld/COFF/Chunks.cpp:416]: Maybe `offset` instead of `padding` here?

**Review by mstorsjo — APPROVED** [review]
LGTM now, thanks! Just one comment wording suggestion.

File: lld/COFF/Chunks.cpp (hunk 10 above)
Comment (cjacek) [reply lld/COFF/Chunks.cpp]: I changed it and pushed. Thanks!
